openstack_openapi/openapi.rst

```diff
@@ -0,0 +1,465 @@
+   described explicitly (name, data type, description)
+
+`Microversion <https://specs.openstack.org/openstack/api-sig/guidelines/microversion_specification.html>`__
+-----------------------------------------------------------------------------------------------------------
+
+“API Microversions” allow changes to the API while preserving backward
+compatibility. The basic idea is that a user has to explicitly ask for their
+request to be treated with a particular version of the API.
+
+So breaking changes can be added to the API without breaking users who don’t
+specifically ask for it.
+
+.. code:: console
+
+   $> curl -i $publicURL/volumes -X POST -H "OpenStack-API-Version: volume 3.70" ...
+
+. . .
+
+.. danger::
+
+   OpenAPI does not support different operations based on the
+   passed headers
+
+RPC style actions
+-----------------
+
+Few OpenStack services rely on the RPC like actions where operation is routed
+based on the body.
+
+.. code:: console
+
+   $> curl -i $publicURL/servers/fake_id/action -X POST -d '
+       {
+           "lock": {"locked_reason": "I don't want to work"}
+       }
+   '
+
+RPC style actions
+-----------------
+
+.. image:: server-actions.png
+
+. . .
+
+.. danger::
+
+   OpenAPI does not support different operations based on the
+   request body
+
+So now what?
+------------
+
+“X-” extensions
+---------------
+
+OpenAPI allows *X-custom* extensions
+
+.. code:: yaml
+
+   paths:
+     /servers:
+       put:
+         ...
+         x-openstack: 
+           foo: bar
+
+
+Dynamic headers
+---------------
+
+Custom OpenAPI parameter serialization based on regex (similar to parameter serialization)
+
+.. code:: yaml
+
+   paths:
+     /foo:
+       put:
+         ...
+         parameters:
+           - in: header
+             name: X-Account-Meta-*
+             schema:
+               type: string
+             x-openstack:
+               style: regex
+
+Microversions
+-------------
+
+JSON schema ``oneOf`` with ``x-openstack`` extension and custom discriminator
+
+.. code:: yaml
+
+   components:
+     schemas:
+       VolumesCreateRequest:
+         oneOf:
+           - $ref: '#/components/schemas/VolumesCreate_30'
+           - $ref: '#/components/schemas/VolumesCreate_313'
+           - $ref: '#/components/schemas/VolumesCreate_347'
+           - $ref: '#/components/schemas/VolumesCreate_353'
+         x-openstack:
+           discriminator: microversion
+       VolumesCreate_313:
+         ...
+         x-openstack:
+           min-ver: 3.13
+           max-ver: 3.47
+       ...
+
+RPC style Actions
+-----------------
+
+JSON schema ``oneOf`` with ``x-openstack`` extension and custom discriminator
+
+.. code:: yaml
+
+   components:
+     schemas:
+       server_actions:
+         oneOf:
+           - $ref: #/components/schemas/action_foo
+           - $ref: #/components/schemas/action_bar
+         x-openstack:
+           discriminator: action
```

Comment:
action here means the differentiator on the body request right?

Reply:
yes, this is a custom discriminator that is understood by dedicated tooling